Feature/multi md steps

iparask opened the pull request:
This PR closes issue #22 and issue #20.

@SrinivasMushnoori, I would like to ask you to give this a try. Before executing please do the following:
1. download fresh RP's and EnTK's repos and checkout `user/repex` branches.
2. create a new virtualenv and install `radical.saga==1.5.7`, `radical.utils==1.5.7` with:
   `pip install radical.utils==1.5.7 radical.saga==1.5.7`
3. Install RP from the `user/repex` branch and EnTK from the `user/repex` branch
4. Install Repex from the `feature/multi_md_steps`

In the end your radical-stack should look like:

```
  radical.entk         : 1.5.5-v1.5.5-44-g597fad2@user-repex
  radical.gtod         : 1.5.0
  radical.pilot        : 1.5.7-v1.5.7-109-g7895d39@user-repex
  radical.repex        : 0.72.0-f75af10@feature-multi_md_steps
  radical.saga         : 1.5.7
  radical.utils        : 1.5.7
```

Under `tests/unittests/test_case` there is a `workflow_gromacs.json`. Please use that as the workflow file as it is the one I use to run. Please use Bridges as your target machine and let me know here how it goes.

andre-merzky commented on src/radical/repex/replica.py:
```diff
@@ -125,26 +128,30 @@ def add_md_stage(self, exchanged_from=None, sid=None, last=False):
-        td['download_output_data'] = copy_outputs
+        # td['link_input_data']      = link_inputs
+        # td['download_output_data'] = copy_outputs
+        for i in range(len(td['description'])):
```
This assumes that the `td['description']` is now always a list, right?  If so, I would suggest we accept dicts as well, and add this before the loop:
```py
for i, descr in enumerate(ru.as_list(td['description'])):
  ...
```
and then use `descr` instead of `td['description'][i]` in the loop body.  That should make this backward compatible (and simpler for the case where we have 'normal' tasks).

andre-merzky commented on src/radical/repex/replica.py:
```diff
@@ -125,26 +128,30 @@ def add_md_stage(self, exchanged_from=None, sid=None, last=False):
-        td['link_input_data']      = link_inputs
-        td['download_output_data'] = copy_outputs
+        # td['link_input_data']      = link_inputs
+        # td['download_output_data'] = copy_outputs
```
These comments can be removed.

andre-merzky commented on src/radical/repex/replica.py:
```diff
@@ -182,20 +189,21 @@ def add_ex_stage(self, exchange_list, ex_alg, sid):
-
-        self._log.debug('%5s add ex: %s', self.rid, task.name)
+        
+        self._log.debug('%5s added ex: %s, input data: %s', self.rid, task.name, task.link_input_data)
```
This might become a bit unreadable?  Not sure, leave it if you think it's ok...

andre-merzky commented on src/radical/repex/utils.py:
```diff
@@ -20,7 +20,10 @@ def expand_ln(to_link, src_sbox, tgt_sbox, rid, cycle):
         except:
             raise RuntimeError('expansion error: %s : %s : %s' % (src, tgt, expand))
-        ret.append('%s/%s > %s/%s' % (src_sbox, src, tgt_sbox, tgt))
+        if task_id:
```
This will fail for `task_id = 0`?  Maybe `if task_id is not None`?

andre-merzky reviewed (commented):
This looks good to me - some small comments, and let's see what @SrinivasMushnoori says :-)  Thanks!

iparask commented on src/radical/repex/exchange.py:
```diff
@@ -76,9 +76,10 @@ def __init__(self, workload, resource, replicas=None):
         rmq_host = str(self._resource.get('rmq_host', 'localhost'))
         rmq_port = int(self._resource.get('rmq_port', '5672'))
-
+        rmq_uname = str(self._resource.get('rmq_uname','guest'))
```
okay

iparask commented:
Hello @SrinivasMushnoori,

I think this branch of repex is ready from my point of view. Can you please give it a try with the following stack
```
  radical.entk         : 1.6.0
  radical.gtod         : 1.5.0
  radical.pilot        : 1.6.2-v1.6.2-156-g9c52e36@user-repex
  radical.repex        : 0.72.0-f395f96@feature-multi_md_steps
  radical.saga         : 1.6.1
  radical.utils        : 1.6.2
```
Let me know here how it goes.

SrinivasMushnoori commented:
On it. Thank you.

iparask commented:
I updated the badges, the event triggers and added more information in the integration test readme file.

I changed the event names and have them unique per resource. I found that all three workflows were triggered by the same event name.

SrinivasMushnoori commented:
seems to be working!

SrinivasMushnoori reviewed (approved):
In `workflow_gromacs.json`, some things need to be changed with respect to the `pre_exec` of each MD task. Specifically, the pre exec needs to be:
```
["module load gromacs/2020.2-cpu",
  "export GMX_MAXBACKUP=-1"]
```

However I am approving since this is not explicitly a RepEx problem, just needs to be fixed in the test case.
